src/properties.rs

```diff
@@ -43,7 +60,7 @@ pub(crate) enum PropertyIdentifier {
 }
 
 /// All of the possible properties that MQTT version 5 supports.
-#[derive(Debug, Copy, Clone, PartialEq)]
+#[derive(Debug, Copy, Clone, PartialEq, Deserialize)]
```

Comment:
One issue that I'm hitting is the fact that we cannot use this `Deserialize` derive macro.

It looks like during deserialization, the deserializer grabs the `PropertyIdentifier` from the message and uses that to figure out which `Property` variant to construct. However, the mapping from `PropertyIdentifier` -> `Property` is not one-to-one, as there are numerous `PropertyIdentifiers` that are unused.

Reply:
The cargo-expand for this `derive` is below:

```rust
    #[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
    const _: () = {
        #[allow(unused_extern_crates, clippy::useless_attribute)]
        extern crate serde as _serde;
        #[automatically_derived]
        impl<'de: 'a, 'a> _serde::Deserialize<'de> for Property<'a> {
            fn deserialize<__D>(
                __deserializer: __D,
            ) -> _serde::__private::Result<Self, __D::Error>
            where
                __D: _serde::Deserializer<'de>,
            {
                #[allow(non_camel_case_types)]
                enum __Field {
                    __field0,
                    __field1,
                    __field2,
                    __field3,
                    __field4,
                    __field5,
                    __field6,
                    __field7,
                    __field8,
                    __field9,
                    __field10,
                    __field11,
                    __field12,
                    __field13,
                    __field14,
                    __field15,
                    __field16,
                    __field17,
                    __field18,
                    __field19,
                    __field20,
                    __field21,
                    __field22,
                    __field23,
                    __field24,
                    __field25,
                    __field26,
                }
                struct __FieldVisitor;
                impl<'de> _serde::de::Visitor<'de> for __FieldVisitor {
                    type Value = __Field;
                    fn expecting(
                        &self,
                        __formatter: &mut _serde::__private::Formatter,
                    ) -> _serde::__private::fmt::Result {
                        _serde::__private::Formatter::write_str(
                            __formatter,
                            "variant identifier",
                        )
                    }
                    fn visit_u64<__E>(
                        self,
                        __value: u64,
                    ) -> _serde::__private::Result<Self::Value, __E>
                    where
                        __E: _serde::de::Error,
                    {
                        match __value {
                            0u64 => _serde::__private::Ok(__Field::__field0),
                            1u64 => _serde::__private::Ok(__Field::__field1),
                            2u64 => _serde::__private::Ok(__Field::__field2),
                            3u64 => _serde::__private::Ok(__Field::__field3),
                            4u64 => _serde::__private::Ok(__Field::__field4),
                            5u64 => _serde::__private::Ok(__Field::__field5),
                            6u64 => _serde::__private::Ok(__Field::__field6),
                            7u64 => _serde::__private::Ok(__Field::__field7),
                            8u64 => _serde::__private::Ok(__Field::__field8),
                            9u64 => _serde::__private::Ok(__Field::__field9),
                            10u64 => _serde::__private::Ok(__Field::__field10),
                            11u64 => _serde::__private::Ok(__Field::__field11),
                            12u64 => _serde::__private::Ok(__Field::__field12),
                            13u64 => _serde::__private::Ok(__Field::__field13),
                            14u64 => _serde::__private::Ok(__Field::__field14),
                            15u64 => _serde::__private::Ok(__Field::__field15),
                            16u64 => _serde::__private::Ok(__Field::__field16),
                            17u64 => _serde::__private::Ok(__Field::__field17),
                            18u64 => _serde::__private::Ok(__Field::__field18),
                            19u64 => _serde::__private::Ok(__Field::__field19),
                            20u64 => _serde::__private::Ok(__Field::__field20),
                            21u64 => _serde::__private::Ok(__Field::__field21),
                            22u64 => _serde::__private::Ok(__Field::__field22),
                            23u64 => _serde::__private::Ok(__Field::__field23),
                            24u64 => _serde::__private::Ok(__Field::__field24),
                            25u64 => _serde::__private::Ok(__Field::__field25),
                            26u64 => _serde::__private::Ok(__Field::__field26),
                            _ => {
                                _serde::__private::Err(
                                    _serde::de::Error::invalid_value(
                                        _serde::de::Unexpected::Unsigned(__value),
                                        &"variant index 0 <= i < 27",
                                    ),
                                )
                            }
                        }
                    }
                    fn visit_str<__E>(
                        self,
                        __value: &str,
                    ) -> _serde::__private::Result<Self::Value, __E>
                    where
                        __E: _serde::de::Error,
                    {
                        match __value {
                            "PayloadFormatIndicator" => {
                                _serde::__private::Ok(__Field::__field0)
                            }
                            "MessageExpiryInterval" => {
                                _serde::__private::Ok(__Field::__field1)
                            }
                            "ContentType" => _serde::__private::Ok(__Field::__field2),
                            "ResponseTopic" => _serde::__private::Ok(__Field::__field3),
                            "CorrelationData" => _serde::__private::Ok(__Field::__field4),
                            "SubscriptionIdentifier" => {
                                _serde::__private::Ok(__Field::__field5)
                            }
                            "SessionExpiryInterval" => {
                                _serde::__private::Ok(__Field::__field6)
                            }
                            "AssignedClientIdentifier" => {
                                _serde::__private::Ok(__Field::__field7)
                            }
                            "ServerKeepAlive" => _serde::__private::Ok(__Field::__field8),
                            "AuthenticationMethod" => {
                                _serde::__private::Ok(__Field::__field9)
                            }
                            "AuthenticationData" => {
                                _serde::__private::Ok(__Field::__field10)
                            }
                            "RequestProblemInformation" => {
                                _serde::__private::Ok(__Field::__field11)
                            }
                            "WillDelayInterval" => {
                                _serde::__private::Ok(__Field::__field12)
                            }
                            "RequestResponseInformation" => {
                                _serde::__private::Ok(__Field::__field13)
                            }
                            "ResponseInformation" => {
                                _serde::__private::Ok(__Field::__field14)
                            }
                            "ServerReference" => {
                                _serde::__private::Ok(__Field::__field15)
                            }
                            "ReasonString" => _serde::__private::Ok(__Field::__field16),
                            "ReceiveMaximum" => _serde::__private::Ok(__Field::__field17),
                            "TopicAliasMaximum" => {
                                _serde::__private::Ok(__Field::__field18)
                            }
                            "TopicAlias" => _serde::__private::Ok(__Field::__field19),
                            "MaximumQoS" => _serde::__private::Ok(__Field::__field20),
                            "RetainAvailable" => {
                                _serde::__private::Ok(__Field::__field21)
                            }
                            "UserProperty" => _serde::__private::Ok(__Field::__field22),
                            "MaximumPacketSize" => {
                                _serde::__private::Ok(__Field::__field23)
                            }
                            "WildcardSubscriptionAvailable" => {
                                _serde::__private::Ok(__Field::__field24)
                            }
                            "SubscriptionIdentifierAvailable" => {
                                _serde::__private::Ok(__Field::__field25)
                            }
                            "SharedSubscriptionAvailable" => {
                                _serde::__private::Ok(__Field::__field26)
                            }
                            _ => {
                                _serde::__private::Err(
                                    _serde::de::Error::unknown_variant(__value, VARIANTS),
                                )
                            }
                        }
                    }
                    fn visit_bytes<__E>(
                        self,
                        __value: &[u8],
                    ) -> _serde::__private::Result<Self::Value, __E>
                    where
                        __E: _serde::de::Error,
                    {
                        match __value {
                            b"PayloadFormatIndicator" => {
                                _serde::__private::Ok(__Field::__field0)
                            }
                            b"MessageExpiryInterval" => {
                                _serde::__private::Ok(__Field::__field1)
                            }
                            b"ContentType" => _serde::__private::Ok(__Field::__field2),
                            b"ResponseTopic" => _serde::__private::Ok(__Field::__field3),
                            b"CorrelationData" => {
                                _serde::__private::Ok(__Field::__field4)
                            }
                            b"SubscriptionIdentifier" => {
                                _serde::__private::Ok(__Field::__field5)
                            }
                            b"SessionExpiryInterval" => {
                                _serde::__private::Ok(__Field::__field6)
                            }
                            b"AssignedClientIdentifier" => {
                                _serde::__private::Ok(__Field::__field7)
                            }
                            b"ServerKeepAlive" => {
                                _serde::__private::Ok(__Field::__field8)
                            }
                            b"AuthenticationMethod" => {
                                _serde::__private::Ok(__Field::__field9)
                            }
                            b"AuthenticationData" => {
                                _serde::__private::Ok(__Field::__field10)
                            }
                            b"RequestProblemInformation" => {
                                _serde::__private::Ok(__Field::__field11)
                            }
                            b"WillDelayInterval" => {
                                _serde::__private::Ok(__Field::__field12)
                            }
                            b"RequestResponseInformation" => {
                                _serde::__private::Ok(__Field::__field13)
                            }
                            b"ResponseInformation" => {
                                _serde::__private::Ok(__Field::__field14)
                            }
                            b"ServerReference" => {
                                _serde::__private::Ok(__Field::__field15)
                            }
                            b"ReasonString" => _serde::__private::Ok(__Field::__field16),
                            b"ReceiveMaximum" => {
                                _serde::__private::Ok(__Field::__field17)
                            }
                            b"TopicAliasMaximum" => {
                                _serde::__private::Ok(__Field::__field18)
                            }
                            b"TopicAlias" => _serde::__private::Ok(__Field::__field19),
                            b"MaximumQoS" => _serde::__private::Ok(__Field::__field20),
                            b"RetainAvailable" => {
                                _serde::__private::Ok(__Field::__field21)
                            }
                            b"UserProperty" => _serde::__private::Ok(__Field::__field22),
                            b"MaximumPacketSize" => {
                                _serde::__private::Ok(__Field::__field23)
                            }
                            b"WildcardSubscriptionAvailable" => {
                                _serde::__private::Ok(__Field::__field24)
                            }
                            b"SubscriptionIdentifierAvailable" => {
                                _serde::__private::Ok(__Field::__field25)
                            }
                            b"SharedSubscriptionAvailable" => {
                                _serde::__private::Ok(__Field::__field26)
                            }
                            _ => {
                                let __value = &_serde::__private::from_utf8_lossy(__value);
                                _serde::__private::Err(
                                    _serde::de::Error::unknown_variant(__value, VARIANTS),
                                )
                            }
                        }
                    }
                }
                impl<'de> _serde::Deserialize<'de> for __Field {
                    #[inline]
                    fn deserialize<__D>(
                        __deserializer: __D,
                    ) -> _serde::__private::Result<Self, __D::Error>
                    where
                        __D: _serde::Deserializer<'de>,
                    {
                        _serde::Deserializer::deserialize_identifier(
                            __deserializer,
                            __FieldVisitor,
                        )
                    }
                }
                struct __Visitor<'de: 'a, 'a> {
                    marker: _serde::__private::PhantomData<Property<'a>>,
                    lifetime: _serde::__private::PhantomData<&'de ()>,
                }
                impl<'de: 'a, 'a> _serde::de::Visitor<'de> for __Visitor<'de, 'a> {
                    type Value = Property<'a>;
                    fn expecting(
                        &self,
                        __formatter: &mut _serde::__private::Formatter,
                    ) -> _serde::__private::fmt::Result {
                        _serde::__private::Formatter::write_str(
                            __formatter,
                            "enum Property",
                        )
                    }
                    fn visit_enum<__A>(
                        self,
                        __data: __A,
                    ) -> _serde::__private::Result<Self::Value, __A::Error>
                    where
                        __A: _serde::de::EnumAccess<'de>,
                    {
                        match match _serde::de::EnumAccess::variant(__data) {
                            _serde::__private::Ok(__val) => __val,
                            _serde::__private::Err(__err) => {
                                return _serde::__private::Err(__err);
                            }
                        } {
                            (__Field::__field0, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::PayloadFormatIndicator,
                                )
                            }
                            (__Field::__field1, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u32,
                                    >(__variant),
                                    Property::MessageExpiryInterval,
                                )
                            }
                            (__Field::__field2, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a str,
                                    >(__variant),
                                    Property::ContentType,
                                )
                            }
                            (__Field::__field3, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a str,
                                    >(__variant),
                                    Property::ResponseTopic,
                                )
                            }
                            (__Field::__field4, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a [u8],
                                    >(__variant),
                                    Property::CorrelationData,
                                )
                            }
                            (__Field::__field5, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u32,
                                    >(__variant),
                                    Property::SubscriptionIdentifier,
                                )
                            }
                            (__Field::__field6, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u32,
                                    >(__variant),
                                    Property::SessionExpiryInterval,
                                )
                            }
                            (__Field::__field7, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a str,
                                    >(__variant),
                                    Property::AssignedClientIdentifier,
                                )
                            }
                            (__Field::__field8, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u16,
                                    >(__variant),
                                    Property::ServerKeepAlive,
                                )
                            }
                            (__Field::__field9, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a str,
                                    >(__variant),
                                    Property::AuthenticationMethod,
                                )
                            }
                            (__Field::__field10, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a [u8],
                                    >(__variant),
                                    Property::AuthenticationData,
                                )
                            }
                            (__Field::__field11, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::RequestProblemInformation,
                                )
                            }
                            (__Field::__field12, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u32,
                                    >(__variant),
                                    Property::WillDelayInterval,
                                )
                            }
                            (__Field::__field13, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::RequestResponseInformation,
                                )
                            }
                            (__Field::__field14, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a str,
                                    >(__variant),
                                    Property::ResponseInformation,
                                )
                            }
                            (__Field::__field15, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a str,
                                    >(__variant),
                                    Property::ServerReference,
                                )
                            }
                            (__Field::__field16, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        &'a str,
                                    >(__variant),
                                    Property::ReasonString,
                                )
                            }
                            (__Field::__field17, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u16,
                                    >(__variant),
                                    Property::ReceiveMaximum,
                                )
                            }
                            (__Field::__field18, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u16,
                                    >(__variant),
                                    Property::TopicAliasMaximum,
                                )
                            }
                            (__Field::__field19, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u16,
                                    >(__variant),
                                    Property::TopicAlias,
                                )
                            }
                            (__Field::__field20, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::MaximumQoS,
                                )
                            }
                            (__Field::__field21, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::RetainAvailable,
                                )
                            }
                            (__Field::__field22, __variant) => {
                                struct __Visitor<'de: 'a, 'a> {
                                    marker: _serde::__private::PhantomData<Property<'a>>,
                                    lifetime: _serde::__private::PhantomData<&'de ()>,
                                }
                                impl<'de: 'a, 'a> _serde::de::Visitor<'de>
                                for __Visitor<'de, 'a> {
                                    type Value = Property<'a>;
                                    fn expecting(
                                        &self,
                                        __formatter: &mut _serde::__private::Formatter,
                                    ) -> _serde::__private::fmt::Result {
                                        _serde::__private::Formatter::write_str(
                                            __formatter,
                                            "tuple variant Property::UserProperty",
                                        )
                                    }
                                    #[inline]
                                    fn visit_seq<__A>(
                                        self,
                                        mut __seq: __A,
                                    ) -> _serde::__private::Result<Self::Value, __A::Error>
                                    where
                                        __A: _serde::de::SeqAccess<'de>,
                                    {
                                        let __field0 = match match _serde::de::SeqAccess::next_element::<
                                            &'a str,
                                        >(&mut __seq) {
                                            _serde::__private::Ok(__val) => __val,
                                            _serde::__private::Err(__err) => {
                                                return _serde::__private::Err(__err);
                                            }
                                        } {
                                            _serde::__private::Some(__value) => __value,
                                            _serde::__private::None => {
                                                return _serde::__private::Err(
                                                    _serde::de::Error::invalid_length(
                                                        0usize,
                                                        &"tuple variant Property::UserProperty with 2 elements",
                                                    ),
                                                );
                                            }
                                        };
                                        let __field1 = match match _serde::de::SeqAccess::next_element::<
                                            &'a str,
                                        >(&mut __seq) {
                                            _serde::__private::Ok(__val) => __val,
                                            _serde::__private::Err(__err) => {
                                                return _serde::__private::Err(__err);
                                            }
                                        } {
                                            _serde::__private::Some(__value) => __value,
                                            _serde::__private::None => {
                                                return _serde::__private::Err(
                                                    _serde::de::Error::invalid_length(
                                                        1usize,
                                                        &"tuple variant Property::UserProperty with 2 elements",
                                                    ),
                                                );
                                            }
                                        };
                                        _serde::__private::Ok(
                                            Property::UserProperty(__field0, __field1),
                                        )
                                    }
                                }
                                _serde::de::VariantAccess::tuple_variant(
                                    __variant,
                                    2usize,
                                    __Visitor {
                                        marker: _serde::__private::PhantomData::<Property<'a>>,
                                        lifetime: _serde::__private::PhantomData,
                                    },
                                )
                            }
                            (__Field::__field23, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<
                                        u32,
                                    >(__variant),
                                    Property::MaximumPacketSize,
                                )
                            }
                            (__Field::__field24, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::WildcardSubscriptionAvailable,
                                )
                            }
                            (__Field::__field25, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::SubscriptionIdentifierAvailable,
                                )
                            }
                            (__Field::__field26, __variant) => {
                                _serde::__private::Result::map(
                                    _serde::de::VariantAccess::newtype_variant::<u8>(__variant),
                                    Property::SharedSubscriptionAvailable,
                                )
                            }
                        }
                    }
                }
                const VARIANTS: &'static [&'static str] = &[
                    "PayloadFormatIndicator",
                    "MessageExpiryInterval",
                    "ContentType",
                    "ResponseTopic",
                    "CorrelationData",
                    "SubscriptionIdentifier",
                    "SessionExpiryInterval",
                    "AssignedClientIdentifier",
                    "ServerKeepAlive",
                    "AuthenticationMethod",
                    "AuthenticationData",
                    "RequestProblemInformation",
                    "WillDelayInterval",
                    "RequestResponseInformation",
                    "ResponseInformation",
                    "ServerReference",
                    "ReasonString",
                    "ReceiveMaximum",
                    "TopicAliasMaximum",
                    "TopicAlias",
                    "MaximumQoS",
                    "RetainAvailable",
                    "UserProperty",
                    "MaximumPacketSize",
                    "WildcardSubscriptionAvailable",
                    "SubscriptionIdentifierAvailable",
                    "SharedSubscriptionAvailable",
                ];
                _serde::Deserializer::deserialize_enum(
                    __deserializer,
                    "Property",
                    VARIANTS,
                    __Visitor {
                        marker: _serde::__private::PhantomData::<Property<'a>>,
                        lifetime: _serde::__private::PhantomData,
                    },
                )
            }
        }
    };
```

Reply:
Can https://docs.rs/num_enum/latest/num_enum/ help here? It tends to provide finer tools for customization of enum discriminants.

Reply:
No, this is a limitation of Rust. You can't assign a non-unit enumeration with an assocaited value. E.g.:
```
enum Foo {
    Bar(u32) = 5,
}
```

is experimental currently.

Reply:
Ack. So at least for `Property` we'd still have to implement deser manually with a big match, right?